tests/bluetooth/host/keys/bt_keys_clear/src/main.c

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/zephyr.h>
+#include <bluetooth/bluetooth.h>
+#include "testing_common_defs.h"
+#include <keys.h>
+#include "host_mocks/assert.h"
+#include "mocks/rpa.h"
+#include "mocks/id.h"
+#include "mocks/id_expects.h"
+#include "mocks/keys_help_utils.h"
+
+/* This LUT contains different combinations of ID, Address and key type.
+ * Item in this list will be used to fill keys pool.
+ */
+static const struct id_addr_pair testing_id_addr_pair_lut[] = {
+
+	{ BT_ADDR_ID_1, BT_ADDR_LE_1 },
+	{ BT_ADDR_ID_1, BT_RPA_ADDR_LE_1 },
+	{ BT_ADDR_ID_1, BT_RPA_ADDR_LE_2 },
+	{ BT_ADDR_ID_1, BT_ADDR_LE_3 },
+
+	{ BT_ADDR_ID_2, BT_ADDR_LE_1 },
+	{ BT_ADDR_ID_2, BT_RPA_ADDR_LE_2 },
+	{ BT_ADDR_ID_2, BT_RPA_ADDR_LE_3 },
+	{ BT_ADDR_ID_2, BT_ADDR_LE_2 },
+
+	{ BT_ADDR_ID_3, BT_ADDR_LE_1 },
+	{ BT_ADDR_ID_3, BT_ADDR_LE_2 },
+
+	{ BT_ADDR_ID_4, BT_ADDR_LE_1 }
+};
+
+/* Global iterator to iterate over the ID, Address and type LUT */
+static uint8_t params_it;
+
+/* Pointer to the current set of testing parameters */
+struct id_addr_pair const *current_params_vector;
+
+/* This list will hold returned references while filling keys pool */
+static struct bt_keys *returned_keys_refs[CONFIG_BT_MAX_PAIRED];
+
+BUILD_ASSERT(ARRAY_SIZE(testing_id_addr_pair_lut) == CONFIG_BT_MAX_PAIRED);
+BUILD_ASSERT(ARRAY_SIZE(testing_id_addr_pair_lut) == ARRAY_SIZE(returned_keys_refs));
+
+/*
+ *  Clear an existing key and verify the result
+ *
+ *  Constraints:
+ *   - Key reference points to a valid item
+ *
+ *  Expected behaviour:
+ *   - The key content is cleared
+ */
+void test_verify_key_cleared_correctly(void)
+{
+	struct bt_keys empty_key;
+	struct bt_keys *key_ref_to_clear, *find_returned_ref;
+	uint8_t id_to_find = current_params_vector->id;
+	const bt_addr_le_t *addr_to_find = current_params_vector->addr;
+
+	if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+		ztest_test_skip();
+	}
+
+	memset(&empty_key, 0x00, sizeof(struct bt_keys));
+	/* 'params_it' is incremented before running the test case */
+	key_ref_to_clear = returned_keys_refs[params_it - 1];
+
+	/* Ensure that item exists in the keys pool */
+	find_returned_ref = bt_keys_find_addr(id_to_find, addr_to_find);
+	zassert_true(find_returned_ref != NULL, "bt_keys_find_addr() returned a NULL reference");
+
+	bt_keys_clear(key_ref_to_clear);
+
+	expect_not_called_bt_id_del();
+
+	/* Verify that memory was cleared */
+	zassert_mem_equal(key_ref_to_clear, &empty_key, sizeof(struct bt_keys),
+			  "Key content wasn't cleared by 'bt_keys_clear()'");
+
+	/* Ensure that item doesn't exist in the keys pool after calling bt_keys_clear() */
+	find_returned_ref = bt_keys_find_addr(id_to_find, addr_to_find);
```

Comment:
If the key was in the key pool but didn't have BT_KEYS_ID_ADDED set in the state, it's presumably not going to be deleted in this case, so shouldn't this always fail is the previous bt_keys_find_addr suceeded?

Reply:
Yes, the second check is checking if the return value is `NULL`

```
/* Ensure that item doesn't exist in the keys pool after calling bt_keys_clear() */
find_returned_ref = bt_keys_find_addr(id_to_find, addr_to_find);
zassert_true(find_returned_ref == NULL,
		     "bt_keys_find_addr() returned a non-NULL reference");
```